Plutarch/Lift.hs

```diff
@@ -125,8 +139,54 @@ instance
   pconstantToRepr = coerce
   pconstantFromRepr = Just . coerce
 
--- TODO: Add haddock
-newtype DerivePConstantViaNewtype (h :: Type) (p :: PType) (p' :: PType) = DerivePConstantViaNewtype h
+{- | Newtype wrapper for deriving @PConstant@ when the wrapped type is represented
+indirectly by a builtin UPLC type that is /not/ @Data@.
+
+  Ex: @PPubKeyHash@ is a newtype to a @PByteString@ and @PByteString@ is directly
+  represented as a builtin bytestring.
+
+Polymorphic types can be derived as follows:
+
+>newtype Foo a = Foo a
+>
+>newtype PFoo a s = PFoo (Term s a)
+>
+>instance forall a. PLift a => PUnsafeLiftDecl (PFoo a) where
+>  type PLifted (PFoo a) = Foo (PLifted a)
+>
+>deriving via
+>  ( DerivePConstantViaNewtype
+>      (Foo a)
+>      (PFoo (PConstanted a))
+>      (PConstanted a)
+>  )
+>  instance
+>    PConstantable a =>
+>    PConstant (Foo a)
+-}
+newtype
+  DerivePConstantViaNewtype
+    (h :: Type)
+    (p :: PType) -- PType to associate with the newtype
+    (p' :: PType) -- Underlying UPLC representation type
+  = -- | The Haskell newtype we are deriving a @PConstant@ instance for
+    DerivePConstantViaNewtype h
+
+{- | Type synonym to simplify deriving of @PConstant@ via @DerivePConstantViaNewtype@.
+
+A newtype @Foo a@ is considered "Constantable" if:
+
+- The wrapped type @a@ has a @PConstant@ instance.
+- The lifted type of @a@ has a @PUnsafeLiftDecl@ instance.
+- There is type equality between @a@ and @PLifted (PConstanted a)@.
+
+These constraints are sufficient to derive a @PConstant@ instance for the newtype.
+
+For deriving @PConstant@ for a wrapped type represented in UPLC as @Data@, see
+@DerivePConstantViaData@.
+-}
+type PConstantable :: Type -> Constraint
+type PConstantable a = (a ~ PLifted (PConstanted a), PConstant a)
```

Comment:
We should really name `PConstant` `PConstantDecl` and call this `PConstant`, otherwise people will accidentally use `PConstant` instead of `PConstantable`.

Reply:
that's a good idea - though it's kind of a big breaking change. I'm all for it though - will add it to the changelog.

Reply:
Done! And also simplified a bunch of usage sites with these type aliases - actually a huge improvement! I'm very happy with how it came out. Thanks again @peter-mlabs!